src/NServiceBus.Core/Persistence/Learning/SagaPersister/SagaStorageFile.cs

```diff
@@ -38,26 +66,13 @@ public static SagaStorageFile Create(Guid sagaId, SagaManifest manifest)
             return new SagaStorageFile(new FileStream(filePath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.ReadWrite, DefaultBufferSize, FileOptions.Asynchronous), manifest);
         }
 
-        public void Dispose()
-        {
-            fileStream.Dispose();
-
-            if (isCompleted)
-            {
-                File.Delete(fileStream.Name);
-            }
-
-            fileStream = null;
-        }
-
         public Task Write(IContainSagaData sagaData)
         {
             fileStream.Position = 0;
 
             return WriteWithinLock((stream, data, meta) =>
             {
-                var serializer = meta.Serializer;
-                serializer.WriteObject(stream, data);
+                serializer.Serialize(jsonWriter, data);
```

Comment:
Looks like the `WriteWithinLock` action could be tweaked again, because `SagaManifest` is no longer used.

Since it seems like the goal of the action is to avoid capturing, could the `MarkAsCompleted` call to `WriteWithinLock` also use the jsonWriter? If so you could replace `FileStream` with `JsonTextWriter`.

However, looking at the `MarkAsCompleted` its using a couple fields anyway (`EmptyBytes`, `lastModificationSeenAt`), so there might be more room to optimize... though how important is that anyway?

Maybe the action should just give up on passing in parameters?